pepr/utilities.py

```diff
@@ -51,3 +55,270 @@ def assign_record_ids_to_target_models(
+        )
+        # -- Save Step 4
+        if save_path is not None:
+            path = save_path + "/reference_hlf.npy"
+            logger.info(f"Save reference high-level features: {path}.")
+            np.save(path, reference_hlf)
+
+        # -- Compute Step 5
+        logger.info("Extract target high-level features.")
+        target_hlf = DirectGmia._extract_hlf(
+            reference_intermediate_models, attack_eval_data
+        )
+        # -- Save Step 5
+        if save_path is not None:
+            path = save_path + "/target_hlf.npy"
+            logger.info(f"Save target high-level features: {path}.")
+            np.save(path, reference_hlf)
+
+        # -- Compute Step 6
+        logger.info(
+            "Compute pairwise distances between reference and target high-level "
+            "features."
+        )
+        hlf_distances = DirectGmia._calc_pairwise_distances(
+            target_hlf,
+            reference_hlf,
+            filter_pars["hlf_metric"],
+        )
+        # -- Save Step 6 (Results from step 3-6)
+        if save_path is not None:
+            path = (
+                save_path
+                + "/pairwise_distances_hlf_"
+                + filter_pars["hlf_metric"]
+                + ".npy"
+            )
+            logger.info(
+                f"Save pairwise distances between reference and target high-level "
+                f"features: {path}."
+            )
+            np.save(path, hlf_distances)
+
+    else:
+        # -- Or Load results from Steps 3-6
+        path = load_pars["pairwise_distances_hlf_" + filter_pars["hlf_metric"]]
+        logger.info(f"Load distance matrix : {path}.")
+        hlf_distances = np.load(path)
+
+    # -- Compute Step 7
+    number_target_records = None
+    if "number_target_records" in filter_pars.keys():
+        number_target_records = filter_pars["number_target_records"]
+    max_search_rounds = 100
+    if "max_search_rounds" in filter_pars.keys():
+        max_search_rounds = filter_pars["max_search_rounds"]
+    neighbor_threshold = 0.5
+    if "neighbor_threshold" in filter_pars.keys():
+        neighbor_threshold = filter_pars["neighbor_threshold"]
+    probability_threshold = 0.1
+    if "probability_threshold" in filter_pars.keys():
+        probability_threshold = filter_pars["probability_threshold"]
+
+    logger.info("Determine potential vulnerable target records.")
+    (
+        target_records,
+        neighbor_threshold,
+        probability_threshold,
+    ) = DirectGmia._select_target_records(
+        len(reference_train_data),
+        len(target_train_data),
+        hlf_distances,
+        neighbor_threshold,
+        probability_threshold,
+        number_target_records,
+        max_search_rounds,
+    )
+
+    # Step 8
+    filtered_target_indices = np.delete(data_conf["evaluation_indices"], target_records)
+
+    return filtered_target_indices, neighbor_threshold, probability_threshold
```

Comment:
Probably we should rename ``filtered_target_indices``to filtered_data_indices``?

Reply:
Renamed in c3a67ed